[codegen] using TORCH_LIBRARY_FRAGMENT for some namespaces

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88229

Summary:
Sometimes we want to extend an existing custom namespace library, instead of creating a new one,
but we don't have a namespace config right now, so we hardcode some custom libraries defined
in pytorch today, i.e. quantized and quantized_decomposed

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace

Summary:
Assuming it's OK to call TORCH_LIBRARY_FRAGMENT even when we are just trying to initialize a new library, instead of extending an existing one.
This is needed since we might be extending a custom (non-aten) namespace

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 commented:
Can you fix the failing test?

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Assuming it's OK to call TORCH_LIBRARY_FRAGMENT even when we are just trying to initialize a new library, instead of extending an existing one.
This is needed since we might be extending a custom (non-aten) namespace

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
> Can you fix the failing test?

done, please take a look again

ezyang commented:
IMO, this should be configurable on a per namespace basis. If custom namespace lives entitely in yaml it is reasonable to want it to be closed.

larryliu0820 commented:
> IMO, this should be configurable on a per namespace basis. If custom namespace lives entitely in yaml it is reasonable to want it to be closed.

I think one way to support this is to add a tag for a function. For example,

```
- func: quantized::add.out(Tensor a, ..., *, Tensor(a!) out) -> Tensor(a!)
  variants: function
  dispatch:
    CPU: quantized_add_out
  tags: fragment
```

This means that we are expecting a `quantized` library already exist in PyTorch (`quantized/library.cpp`) and just extending it by adding a new `add.out` operator.

jerryzh168 commented:
> > IMO, this should be configurable on a per namespace basis. If custom namespace lives entitely in yaml it is reasonable to want it to be closed.
> 
> I think one way to support this is to add a tag for a function. For example,
> 
> ```
> - func: quantized::add.out(Tensor a, ..., *, Tensor(a!) out) -> Tensor(a!)
>   variants: function
>   dispatch:
>     CPU: quantized_add_out
>   tags: fragment
> ```
> 
> This means that we are expecting a `quantized` library already exist in PyTorch (`quantized/library.cpp`) and just extending it by adding a new `add.out` operator.

should this config be library specific instead of operator specific?

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Assuming it's OK to call TORCH_LIBRARY_FRAGMENT even when we are just trying to initialize a new library, instead of extending an existing one.
This is needed since we might be extending a custom (non-aten) namespace

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 commented on torchgen/gen.py:
```diff
@@ -1623,8 +1623,11 @@ def get_native_function_schema_registrations(
+    ns_to_is_fragment: Dict[str, bool] = defaultdict(bool)
     for native_function in native_functions:
         ns_native_functions[native_function.namespace].append(native_function)
+        ns_to_is_fragment[native_function.namespace] |= "fragment" in native_function.tags
```
Add it here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/tags.yaml

larryliu0820 commented on tools/test/test_codegen.py:
```diff
@@ -217,6 +217,11 @@ def setUp(self) -> None:
             loc=torchgen.model.Location(__file__, 1),
             valid_tags=set(),
         )
+        self.fragment_custom_native_function, _ = torchgen.model.NativeFunction.from_yaml(
+            {"func": "custom::func() -> bool"},
+            loc=torchgen.model.Location(__file__, 1),
+            valid_tags=set("fragment"),
+        )
```
Other than that can you add a test to parse `tags: fragment` from yaml file?

jerryzh168 replied:
OK, sounds good

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Assuming it's OK to call TORCH_LIBRARY_FRAGMENT even when we are just trying to initialize a new library, instead of extending an existing one.
This is needed since we might be extending a custom (non-aten) namespace

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Assuming it's OK to call TORCH_LIBRARY_FRAGMENT even when we are just trying to initialize a new library, instead of extending an existing one.
This is needed since we might be extending a custom (non-aten) namespace

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented:
The config should not live on an op. Should be somewhere library info.

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Sometimes we want to extend an existing custom namespace library, instead of creating a new one,
but we don't have a namespace config right now, so we hardcode some custom libraries defined
in pytorch today, i.e. quantized and quantized_decomposed

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Sometimes we want to extend an existing custom namespace library, instead of creating a new one,
but we don't have a namespace config right now, so we hardcode some custom libraries defined
in pytorch today, i.e. quantized and quantized_decomposed

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented on torchgen/gen.py:
```diff
@@ -1640,8 +1640,16 @@ def get_native_function_schema_registrations(
         else:
             custom_namespace = namespace
             tab = "\t"
+            PREDEFINED_NAMESPACES = set(["quantized", "quantized_decomposed"])
```
You can use curly here to make a set literal

ezyang replied:
Preferably put this top level and better name eg FRAGMENT_NAMESPACES

larryliu0820 replied:
Also can we add it to https://github.com/pytorch/pytorch/blob/99c07735e457a2961f2319b4ba19f0d04eb47967/torchgen/model.py#L49

ezyang reviewed (approved):
I did not review tests

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Sometimes we want to extend an existing custom namespace library, instead of creating a new one,
but we don't have a namespace config right now, so we hardcode some custom libraries defined
in pytorch today, i.e. quantized and quantized_decomposed

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[codegen] Changing TORCH_LIBRARY to TORCH_LIBRARY_FRAGMENT for custom namespace"


Summary:
Sometimes we want to extend an existing custom namespace library, instead of creating a new one,
but we don't have a namespace config right now, so we hardcode some custom libraries defined
in pytorch today, i.e. quantized and quantized_decomposed

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g

jerryzh168 commented:
@pytorchbot merge -f

jerryzh168 commented:
@pytorchbot merge -f "failing tests unrelated"
